### PR TITLE
rgw: add 'bucket sync checkpoint' command to radosgw-admin

### PR DESCRIPTION
**Pull request by cbodley**
Adds a consistency checkpoint command for bucket sync. This fetches the state of the bucket's sources from a remote zone, then polls the local zone's sync status until it catches up.

This command is then used directly in our multisite tests to implement `zone_bucket_checkpoint()`.

```
[zone a1] $ radosgw-admin bucket sync checkpoint --bucket testbucket --source-zone a2
2020-08-05T13:44:00.975-0400 7fcb766b5980  1 waiting for incremental sync to catch up:
      local status: [00000000004.8.5]
    remote markers: [00000000008.18.5]
2020-08-05T13:44:05.977-0400 7fcb766b5980  1 waiting for incremental sync to catch up:
      local status: [00000000004.8.5]
    remote markers: [00000000008.18.5]
2020-08-05T13:44:10.980-0400 7fcb766b5980  1 waiting for incremental sync to catch up:
      local status: [00000000004.8.5]
    remote markers: [00000000008.18.5]
2020-08-05T13:44:15.982-0400 7fcb766b5980  1 bucket sync caught up with source:
      local status: [00000000008.18.5]
    remote markers: [00000000008.18.5]
2020-08-05T13:44:15.982-0400 7fcb766b5980  0 bucket checkpoint complete
```

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_admin.cc [hunk 1]
```diff
@@ -2639,6 +2642,234 @@ static int bucket_sync_status(rgw::sal::RGWRadosStore *store, const RGWBucketInf
...
+  }
+}
+
+bool operator<(const std::vector<rgw_bucket_shard_sync_info>& lhs,
+               const BucketIndexShardsManager& rhs)
+{
+  for (size_t i = 0; i < lhs.size(); ++i) {
+    const auto& l = BucketIndexShardsManager::get_shard_marker(lhs[i].inc_marker.position);
+    const auto& r = rhs.get(i, "");
+    if (l >= r) {
```
Comment (yehudasa) [inline src/rgw/rgw_admin.cc:2672]: @cbodley I assume we have a few shards that are greater or equal and a others that are smaller, so maybe the logic needs to be reversed? That is if `(l < r)` return `true`.

File: src/rgw/rgw_admin.cc [hunk 2]
```diff
@@ -2639,6 +2642,234 @@ static int bucket_sync_status(rgw::sal::RGWRadosStore *store, const RGWBucketInf
...
+      rgw_bucket source_bucket;
+      r = init_bucket(*pipe.source.bucket, source_bucket_info, source_bucket);
+      if (r < 0) {
+        lderr(store->ctx()) << "failed to read source bucket info: " << cpp_strerror(r) << dendl;
+        return r;
+      }
+
+      pipe.source.bucket = source_bucket;
+      pipe.dest.bucket = info.bucket;
+      r = bucket_source_sync_checkpoint(store, zone, z->second, c->second,
```
Comment (yehudasa) [inline src/rgw/rgw_admin.cc:2860]: @cbodley I'd split the checkpoint into two: first get all the remote markers of all sources, and only then wait. This way we reduce the wait on stuff that arrived when we were waiting on another source.

**Comment by cbodley** [comment]
updates:
* moved checkpoint code into new `rgw_sync_checkpoint.*`
* fixed marker comparison logic
* enabled optional_yield for `remote_bilog_markers()`
* initial steps (fetching remote markers and reading bucket info for each source) are done in parallel using optional_yield

**Comment by adamemerson** [comment]
Jenkins, recheck this please

**Comment by adamemerson** [comment]
This needs a fix to pass run-cli-tests
```
diff --git a/src/test/cli/radosgw-admin/help.t b/src/test/cli/radosgw-admin/help.t
index 220d5da7b0c..3a868d7d4ba 100644
--- a/src/test/cli/radosgw-admin/help.t
+++ b/src/test/cli/radosgw-admin/help.t
@@ -29,6 +29,7 @@
     bucket chown               link bucket to specified user and update its object ACLs
     bucket reshard             reshard bucket
     bucket rewrite             rewrite all objects in the specified bucket
+    bucket sync checkpoint     poll a bucket's sync status until it catches up to its remote
     bucket sync disable        disable bucket sync
     bucket sync enable         enable bucket sync
     bucket radoslist           list rados objects backing bucket's objects
```

**Comment by adamemerson** [comment]
It looks like Bucket Creation Time is still failing on Teuthology.

http://qa-proxy.ceph.com/teuthology/aemerson-2020-08-28_05:57:46-rgw-wip-aemerson-qa-distro-basic-smithi/5381836/teuthology.log

**Comment by cbodley** [comment]
thanks @adamemerson! i'll push an update for `help.t`. the `test_bucket_creation_time` fix in https://github.com/ceph/ceph/pull/36743 hasn't merged yet

**Comment by cbodley** [comment]
rebased and amended help.t accordingly

**Comment by cbodley** [comment]
https://pulpito.ceph.com/cbodley-2020-09-03_19:12:11-rgw:multisite-wip-cbodley-testing-distro-basic-smithi/

**Comment by cbodley** [comment]
this is ready for another round of review!

**Comment by cbodley** [comment]
jenkins test api

**Review by adamemerson — APPROVED** [review]

**Pull request merged**